server/src/main/kotlin/com/trib3/server/graphql/GraphQLWebSocket.kt

```diff
@@ -6,15 +6,109 @@ import com.trib3.server.filters.RequestIdFilter
 import graphql.ExecutionInput
 import graphql.ExecutionResult
 import graphql.GraphQL
+import io.reactivex.BackpressureStrategy
+import io.reactivex.Emitter
+import io.reactivex.Flowable
+import io.reactivex.FlowableEmitter
+import io.reactivex.FlowableOnSubscribe
+import io.reactivex.Observable
+import io.reactivex.Scheduler
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
 import mu.KotlinLogging
 import org.eclipse.jetty.websocket.api.StatusCode
 import org.eclipse.jetty.websocket.api.WebSocketAdapter
 import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import java.util.concurrent.TimeUnit
 import java.util.concurrent.locks.ReentrantLock
 import kotlin.concurrent.withLock
 
 private val log = KotlinLogging.logger { }
 
+/**
+ * Class that stores the success and failure callbacks for sending keepalive messages to the client
+ */
+internal class KeepAliveCallbacks(private val socket: GraphQLWebSocket, private val messageId: String?) {
+    fun onInterval(count: Long) {
+        log.trace("Sent $count keepalive messages")
+        socket.sendMessage(OperationType.GQL_CONNECTION_KEEP_ALIVE, messageId)
+    }
+
+    fun onError(e: Throwable) {
+        log.error("Error in keepalive subscriber ${e.message}", e)
+    }
+}
+
+/**
+ * [Subscriber] that forwards all events received to its [emitter]
+ */
+private class EmitterSubscriber<T>(
+    private val emitter: Emitter<T>
+) : Subscriber<T> {
+    private lateinit var subscription: Subscription
+
+    override fun onComplete() {
+        emitter.onComplete()
+    }
+
+    override fun onSubscribe(s: Subscription) {
+        this.subscription = s
+        s.request(1)
+    }
+
+    override fun onNext(t: T) {
+        emitter.onNext(t)
+        subscription.request(1)
+    }
+
+    override fun onError(t: Throwable) {
+        emitter.onError(t)
+    }
+}
+
+/**
+ * Class that executes a graphQL query and can be subscribed to to receive
+ * results
+ */
+private class GraphQLResultPublisher(
+    private val graphQL: GraphQL,
+    private val payload: GraphRequest,
+    private val loggingRequestId: String
+) : FlowableOnSubscribe<ExecutionResult> {
+    override fun subscribe(emitter: FlowableEmitter<ExecutionResult>) {
+        RequestIdFilter.withRequestId(loggingRequestId) {
+            try {
+                val result = graphQL.execute(
+                    ExecutionInput.newExecutionInput()
+                        .query(payload.query)
+                        .variables(payload.variables ?: mapOf())
+                        .operationName(payload.operationName)
+                        .build()
+                )
+                // check if the graphQL result is a Publisher
+                val publisherData = try {
```

Comment:
whoa

Reply:
yeah, using control flow statements (try, if, etc) as expressions is really nice.  don't love having to do this via exception, but that looks like the only way the ExecutionResult provides to get/check the data instance type